### ISSUE TITLE
sanity_check_prefix broken/not being called?

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#3746]
I have a WIP package I'm writing and `spack install` just ran successfully. However, nothing was installed! The installation prefix only contains the `.spack` directory. The package extends `MakefilePackage`, which does indeed contain:
```python
# Check that self.prefix is there after installation                        
run_after('install')(PackageBase.sanity_check_prefix)
```
so it should be run. Any ideas how to test this?

**Comment by adamjstewart** [comment]
I don't remember which package this was, but I'm pretty sure I know what the problem was.

If you override the default `phases` for a package base class, all of the `run_before`/`run_after` decorators will be removed, so Spack will no longer run `sanity_check_prefix`, even if `install` is one of your phases. It would be nice if there was a way to fix this.

**Comment by hartzell** [comment]
Can one of you say a bit more about this?  I'm working on a package definition that's basically [this one](https://github.com/spack/spack/pull/11094#issuecomment-481502216) and my run_after/before's never seem to be called (details [here](https://github.com/spack/spack/pull/11094#issuecomment-481504823)). 

I can't figure out why....

**Comment by hartzell** [comment]
Actually, never mind.  It seems that `tty.warn()` and *etc* statements don't work in something being run by `@run_after`, but `sys.exit(42)` certainly does what you'd expect.

**Comment by adamjstewart** [comment]
@hartzell Yes, tty is currently captured during installation, see #2566. This will be fixed once #10412 is complete.

**Comment by hartzell** [comment]
Just filed #11244 with another approach.

**Comment by adamjstewart** [comment]
I haven't encountered this problem in a long time, so I'm going to close this issue.
